Use focus/blur to prevent prop updates from changing `FormText` values while typing

ahmedhamidawan opened the pull request:
We have a bug where in the workflow editor, while typing in a text form element, the value can get unexpectedly trimmed or reset because of the `build_module` API call updating the values. Using a combination of blur and focus to prevent prop values from replacing currently typed values is a solution to this.

Fixes https://github.com/galaxyproject/galaxy/issues/21205

| Before 🔴 |
| ---- |
| <video src="https://github.com/user-attachments/assets/99bffde5-9b34-41ac-a954-bd60808d3acf" /> |

| After 🟢 _(out of date: no longer using a debounce)_ |
| ---- |
| <video src="https://github.com/user-attachments/assets/46754df8-c634-44d6-9cd0-6be423d7f06b" /> |

## How to test the changes?
(Select all options that apply)
- [ ] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [x] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

mvdbeek commented:
The test errors look legitimate to me. Maybe debounce is not the right thing? it seems that other fields are fine, is that maybe because `LastQueue` is used to set those values ?

ahmedhamidawan commented:
I tried a slightly different approach thought it is still via a debounce. Seleniums are still failing (though fewer fails than [earlier](https://github.com/galaxyproject/galaxy/actions/runs/19412288389/job/55535500795?pr=21300)) and _possibly_ related? Will investigate the fails to confirm.

ahmedhamidawan committed:
[25.1] Use focus/blur to prevent prop updates from changing `FormText` values while typing

Fixes https://github.com/galaxyproject/galaxy/issues/21205

ahmedhamidawan commented:
Used a strategy that seems to work well!: Used a combination of blur and focus to prevent prop values from replacing currently typed values (if currently in the focus state, wait for the user to complete typing before replacing with the prop value coming from the backend -> which is eventually replaced by the final typed value if needed).

mvdbeek commented on client/src/components/Form/Elements/FormText.vue:
```diff
@@ -118,6 +125,45 @@ export default {
+                    return "";
+                }
+                return this.multiple
+                    ? val.reduce((str_value, item) => str_value + String(item) + "\n", "")
```
this is (and was) broken btw, it'll add newlines as you type.
```suggestion
                    ? val.join("\n"): String(val[0]);
```

ahmedhamidawan committed:
prevent adding newlines as you type

Co-authored-by: Marius van den Beek <m.vandenbeek@gmail.com>

ahmedhamidawan committed:
add debounce to `onColumnDefinitions` in `FormInputCollection` to prevent overwrite while typing

ahmedhamidawan committed:
remove trailing newline from jest

ahmedhamidawan committed:
remove trailing newline from selenium as well

This is because in the prior implementation we would add a needless
newline at the end of a `multiple = true` `FormText` component

mvdbeek commented on client/src/components/Workflow/Editor/Forms/FormInputCollection.vue:
```diff
@@ -87,11 +87,21 @@ const isRecordType = computed(() => {
-    console.log(newColumnDefinitions);
-    state.column_definitions = newColumnDefinitions;
-    emit("onChange", state);
+    // If existing timer, clear it to prevent emitting the value while the user is still typing
```
(and it is passing through LastQueue, but the behavior also doesn't really make sense)

mvdbeek commented:
I am not convinced this isn't introducing state bugs, but it also seems correct. If you target dev I think we can merge this
